File: src/main/java/com/enterprise/iqk/agent/harness/WorkspaceRuntime.java
package com.enterprise.iqk.agent.harness;

import com.enterprise.iqk.config.properties.AgentHarnessProperties;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.stream.Stream;

@Component
public class WorkspaceRuntime implements AgentRuntime {
    private static final Set<String> SUPPORTED_ACTIONS = Set.of(
            "workspace_list_files",
            "workspace_read_file",
            "workspace_search_text",
            "workspace_propose_patch",
            "workspace_apply_patch",
            "workspace_run_shell"
    );

    private final AgentHarnessProperties harnessProperties;
    private final UnifiedDiffService diffService;
    private final Path workspaceRoot;

    @Autowired
    public WorkspaceRuntime(AgentHarnessProperties harnessProperties, UnifiedDiffService diffService) {
        this(harnessProperties, diffService, Path.of(harnessProperties.getWorkspace().getRoot()));
    }

    WorkspaceRuntime(Path workspaceRoot) {
        this(defaultProperties(workspaceRoot), new UnifiedDiffService(), workspaceRoot);
    }

    WorkspaceRuntime(AgentHarnessProperties harnessProperties, UnifiedDiffService diffService, Path workspaceRoot) {
        this.harnessProperties = harnessProperties;
        this.diffService = diffService;
        this.workspaceRoot = workspaceRoot.toAbsolutePath().normalize();
    }

    @Override
    public String source() {
        return "workspace";
    }

    @Override
    public boolean supports(String action) {
        return SUPPORTED_ACTIONS.contains(action);
    }

    @Override
    public AgentObservation execute(AgentAction action) {
        long startedNs = System.nanoTime();
        try {
            Map<String, Object> payload = switch (action.action()) {
                case "workspace_list_files" -> listFiles(action.actionInput());
                case "workspace_read_file" -> readFile(action.actionInput());
                case "workspace_search_text" -> searchText(action.actionInput());
                case "workspace_propose_patch" -> proposePatch(action.actionInput());
                case "workspace_apply_patch" -> applyPatch(action.actionInput());
                case "workspace_run_shell" -> runCommand(action.actionInput());
                default -> Map.of("status", "error", "message", "unsupported action: " + action.action());
            };
            if ("error".equals(payload.get("status"))) {
                Object message = payload.get("message");
                return AgentObservation.error(source(), String.valueOf(message), elapsedMs(startedNs));
            }
            return AgentObservation.success(source(), payload, elapsedMs(startedNs));
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            return AgentObservation.error(source(), "workspace action failed: " + ex.getMessage(), elapsedMs(startedNs));
        } catch (IOException ex) {
            return AgentObservation.error(source(), "workspace action failed: " + ex.getMessage(), elapsedMs(startedNs));
        } catch (RuntimeException ex) {
            return AgentObservation.error(source(), "workspace action failed: " + ex.getMessage(), elapsedMs(startedNs));
        }
    }

    private Map<String, Object> listFiles(Map<String, Object> input) throws IOException {
        Path root = resolvePath(stringVal(input, "path", "."));
        int maxDepth = Math.min(Math.max(intVal(input.get("maxDepth"), 2), 0), 5);
        List<Map<String, Object>> files = new ArrayList<>();
        try (Stream<Path> stream = Files.walk(root, maxDepth)) {
            stream.sorted(Comparator.naturalOrder())
                    .limit(200)
                    .forEach(path -> files.add(fileSummary(path)));
        }
        return Map.of("root", relative(root), "files", files);
    }

    private Map<String, Object> readFile(Map<String, Object> input) throws IOException {
        Path path = resolvePath(stringVal(input, "path", ""));
        if (!Files.isRegularFile(path)) {
            return Map.of("status", "error", "message", "file not found: " + relative(path));
        }
        int maxFileBytes = Math.max(1, harnessProperties.getWorkspace().getMaxFileBytes());
        int maxBytes = Math.min(Math.max(intVal(input.get("maxBytes"), maxFileBytes), 1), maxFileBytes);
        byte[] bytes = readPrefix(path, maxBytes);
        return Map.of(
                "path", relative(path),
                "content", new String(bytes, StandardCharsets.UTF_8),
                "truncated", Files.size(path) > bytes.length
        );
    }

    private Map<String, Object> searchText(Map<String, Object> input) throws IOException {
        String query = stringVal(input, "query", "");
        Path root = resolvePath(stringVal(input, "path", "."));
        int maxMatches = Math.min(Math.max(intVal(input.get("maxMatches"), 50), 1), 100);
        List<Map<String, Object>> matches = new ArrayList<>();
        try (Stream<Path> stream = Files.walk(root, 8)) {
            List<Path> candidates = stream.filter(Files::isRegularFile)
                    .limit(Math.max(1, harnessProperties.getWorkspace().getMaxSearchFiles()))
                    .toList();
            for (Path path : candidates) {
                if (matches.size() >= maxMatches || Files.size(path) > 1_000_000) {
                    continue;
                }
                List<String> lines;
                try {
                    lines = Files.readAllLines(path, StandardCharsets.UTF_8);
                } catch (IOException ignored) {
                    continue;
                }
                for (int i = 0; i < lines.size() && matches.size() < maxMatches; i++) {
                    String line = lines.get(i);
                    if (line.contains(query)) {
                        matches.add(Map.of(
                                "path", relative(path),
                                "lineNumber", i + 1,
                                "line", line
                        ));
                    }
                }
            }
        }
        return Map.of("query", query, "matches", matches);
    }

    private Map<String, Object> proposePatch(Map<String, Object> input) throws IOException {
        Path path = resolvePath(stringVal(input, "path", ""));
        String content = stringVal(input, "content", "");
        String patch = stringVal(input, "patch", "");
        String oldContent = Files.exists(path) && Files.isRegularFile(path)
                ? Files.readString(path, StandardCharsets.UTF_8)
                : "";
        String diff = StringUtils.hasText(patch)
                ? patch
                : diffService.create(relative(path), oldContent, content);
        Map<String, Object> proposal = new LinkedHashMap<>();
        proposal.put("path", relative(path));
        proposal.put("summary", stringVal(input, "summary", ""));
        proposal.put("contentBytes", content.getBytes(StandardCharsets.UTF_8).length);
        proposal.put("patch", diff);
        proposal.put("wouldCreate", !Files.exists(path));
        proposal.put("applyAction", "workspace_apply_patch");
        return proposal;
    }

    private Map<String, Object> applyPatch(Map<String, Object> input) throws IOException {
        if (!harnessProperties.getWorkspace().isWriteEnabled()) {
            return Map.of("status", "error", "message", "workspace writes are disabled");
        }
        Path path = resolvePath(stringVal(input, "path", ""));
        String content = stringVal(input, "content", "");
        String patch = stringVal(input, "patch", "");
        String nextContent = content;
        if (StringUtils.hasText(patch)) {
            String oldContent = Files.exists(path) && Files.isRegularFile(path)
                    ? Files.readString(path, StandardCharsets.UTF_8)
                    : "";
            nextContent = diffService.apply(oldContent, patch);
        }
        Path parent = path.getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }
        Files.writeString(path, nextContent, StandardCharsets.UTF_8,
                StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
        return Map.of(
                "status", "written",
                "path", relative(path),
                "bytes", nextContent.getBytes(StandardCharsets.UTF_8).length
        );
    }

    private Map<String, Object> runCommand(Map<String, Object> input) throws IOException, InterruptedException {
        if (!harnessProperties.getWorkspace().isShellEnabled()) {
            return Map.of("status", "error", "message", "workspace shell is disabled");
        }
        List<String> command = tokenizeCommand(stringVal(input, "command", ""));
        if (!isAllowedCommand(command)) {
            return Map.of("status", "error", "message", "command is not allowed");
        }
        int defaultTimeout = Math.max(1, harnessProperties.getWorkspace().getCommandTimeoutSeconds());
        int timeoutSeconds = Math.min(Math.max(intVal(input.get("timeoutSeconds"), defaultTimeout), 1), 30);
        ProcessBuilder builder = new ProcessBuilder(command);
        builder.directory(workspaceRoot.toFile());
        Process process = builder.start();
        CompletableFuture<ProcessOutput> stdout = CompletableFuture.supplyAsync(
                () -> readProcessOutput(process.getInputStream()));
        CompletableFuture<ProcessOutput> stderr = CompletableFuture.supplyAsync(
                () -> readProcessOutput(process.getErrorStream()));
        boolean finished = process.waitFor(timeoutSeconds, TimeUnit.SECONDS);
        if (!finished) {
            process.destroyForcibly();
            return Map.of("status", "error", "message", "command timed out");
        }
        ProcessOutput stdoutOutput = stdout.join();
        ProcessOutput stderrOutput = stderr.join();
        String stdoutText = new String(stdoutOutput.bytes(), StandardCharsets.UTF_8);
        String stderrText = new String(stderrOutput.bytes(), StandardCharsets.UTF_8);
        return Map.of(
                "exitCode", process.exitValue(),
                "stdout", stdoutText,
                "stderr", stderrText,
                "output", stdoutText + stderrText,
                "truncated", stdoutOutput.truncated() || stderrOutput.truncated()
        );
    }

    private Map<String, Object> fileSummary(Path path) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("path", relative(path));
        result.put("type", Files.isDirectory(path) ? "directory" : "file");
        if (Files.isRegularFile(path)) {
            try {
                result.put("size", Files.size(path));
            } catch (IOException ignored) {
                result.put("size", null);
            }
        }
        return result;
    }

    private Path resolvePath(String raw) {
        Path resolved = workspaceRoot.resolve(StringUtils.hasText(raw) ? raw : ".").normalize();
        if (!resolved.startsWith(workspaceRoot)) {
            throw new IllegalArgumentException("path escapes workspace root");
        }
        return resolved;
    }

    private String relative(Path path) {
        Path normalized = path.toAbsolutePath().normalize();
        if (normalized.equals(workspaceRoot)) {
            return ".";
        }
        return workspaceRoot.relativize(normalized).toString();
    }

    private byte[] readPrefix(Path path, int maxBytes) throws IOException {
        try (var input = Files.newInputStream(path)) {
            return input.readNBytes(maxBytes);
        }
    }

    private ProcessOutput readProcessOutput(InputStream stream) {
        try (var input = stream) {
            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
            input.transferTo(buffer);
            byte[] raw = buffer.toByteArray();
            int maxOutputBytes = maxCommandOutputBytes();
            if (raw.length <= maxOutputBytes) {
                return new ProcessOutput(raw, false);
            }
            return new ProcessOutput(Arrays.copyOf(raw, maxOutputBytes), true);
        } catch (IOException ex) {
            return new ProcessOutput(("failed to read process output: " + ex.getMessage())
                    .getBytes(StandardCharsets.UTF_8), false);
        }
    }

    private List<String> tokenizeCommand(String command) {
        if (!StringUtils.hasText(command)) {
            return List.of();
        }
        return Stream.of(command.trim().split("\\s+")).filter(StringUtils::hasText).toList();
    }

    private boolean isAllowedCommand(List<String> command) {
        if (command.isEmpty()) {
            return false;
        }
        String executable = command.get(0);
        if (!harnessProperties.getWorkspace().getAllowedCommands().contains(executable)) {
            return false;
        }
        if ("pwd".equals(executable)) {
            return command.size() == 1;
        }
        if ("ls".equals(executable) || "rg".equals(executable)) {
            return true;
        }
        if ("git".equals(executable)) {
            return command.size() >= 2
                    && harnessProperties.getWorkspace().getAllowedGitSubcommands().contains(command.get(1));
        }
        if ("mvn".equals(executable)) {
            return command.stream().anyMatch("test"::equals)
                    && command.stream().allMatch(token -> "mvn".equals(token)
                    || "test".equals(token)
                    || "-q".equals(token)
                    || token.startsWith("-D"));
        }
        return false;
    }

    private int maxCommandOutputBytes() {
        return Math.max(1, harnessProperties.getWorkspace().getMaxCommandOutputBytes());
    }

    private String stringVal(Map<String, Object> input, String key, String fallback) {
        Object raw = input.get(key);
        if (raw == null) {
            return fallback;
        }
        String value = String.valueOf(raw).trim();
        return StringUtils.hasText(value) ? value : fallback;
    }

    private int intVal(Object raw, int fallback) {
        if (raw == null) {
            return fallback;
        }
        try {
            return Integer.parseInt(String.valueOf(raw));
        } catch (Exception ignored) {
            return fallback;
        }
    }

    private long elapsedMs(long startedNs) {
        return TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startedNs);
    }

    private static AgentHarnessProperties defaultProperties(Path workspaceRoot) {
        AgentHarnessProperties properties = new AgentHarnessProperties();
        properties.getWorkspace().setRoot(workspaceRoot.toString());
        properties.getWorkspace().setWriteEnabled(true);
        properties.getWorkspace().setShellEnabled(true);
        return properties;
    }

    private record ProcessOutput(byte[] bytes, boolean truncated) {
    }
}

File: python/src/knowledgeops_py/infrastructure/workspace_runtime.py
"""Bounded local workspace runtime used by trusted Harness actions."""

from __future__ import annotations

import difflib

# Shell execution is constrained to a validated argv allowlist below.
import subprocess  # nosec B404
from dataclasses import dataclass
from pathlib import Path
from time import perf_counter
from typing import Any


@dataclass(frozen=True, slots=True)
class WorkspaceRuntime:
    root: Path
    write_enabled: bool
    shell_enabled: bool
    command_timeout_seconds: int = 10
    max_command_output_bytes: int = 12_000
    max_file_bytes: int = 20_000
    max_search_files: int = 1_000
    allowed_commands: tuple[str, ...] = ("pwd", "ls", "rg", "git", "mvn")
    allowed_git_subcommands: tuple[str, ...] = ("status", "diff", "show", "log", "rev-parse", "branch")

    def __post_init__(self) -> None:
        object.__setattr__(self, "root", self.root.resolve())

    def execute(self, action: str, action_input: dict[str, Any]) -> dict[str, Any]:
        started = perf_counter()
        try:
            payload = self._execute(action, action_input)
            status = str(payload.pop("status", "success"))
            result = {"status": status, **payload, "source": "workspace", "latencyMs": elapsed_ms(started)}
            if status == "error":
                result["message"] = str(payload.get("message") or "workspace action failed")
            return result
        except (OSError, ValueError) as exc:
            return {
                "status": "error",
                "source": "workspace",
                "latencyMs": elapsed_ms(started),
                "message": f"workspace action failed: {exc}",
            }

    def _execute(self, action: str, action_input: dict[str, Any]) -> dict[str, Any]:
        if action == "workspace_list_files":
            return self._list_files(action_input)
        if action == "workspace_read_file":
            return self._read_file(action_input)
        if action == "workspace_search_text":
            return self._search_text(action_input)
        if action == "workspace_propose_patch":
            return self._propose_patch(action_input)
        if action == "workspace_apply_patch":
            return self._apply_patch(action_input)
        if action == "workspace_run_shell":
            return self._run_shell(action_input)
        return {"status": "error", "message": f"unsupported action: {action}"}

    def _list_files(self, action_input: dict[str, Any]) -> dict[str, Any]:
        root = self._resolve_path(string_value(action_input, "path", "."))
        max_depth = min(max(integer_value(action_input.get("maxDepth"), 2), 0), 5)
        files = [self._file_summary(root)]
        for candidate in sorted(root.rglob("*")):
            if len(candidate.relative_to(root).parts) > max_depth:
                continue
            files.append(self._file_summary(candidate))
            if len(files) >= 200:
                break
        return {"root": self._relative(root), "files": files}

    def _read_file(self, action_input: dict[str, Any]) -> dict[str, Any]:
        path = self._resolve_path(string_value(action_input, "path", ""))
        if not path.is_file():
            return {"status": "error", "message": f"file not found: {self._relative(path)}"}
        max_bytes = min(max(integer_value(action_input.get("maxBytes"), self.max_file_bytes), 1), self.max_file_bytes)
        content = path.read_bytes()[:max_bytes]
        return {
            "path": self._relative(path),
            "content": content.decode("utf-8", errors="replace"),
            "truncated": path.stat().st_size > len(content),
        }

    def _search_text(self, action_input: dict[str, Any]) -> dict[str, Any]:
        query = string_value(action_input, "query", "")
        root = self._resolve_path(string_value(action_input, "path", "."))
        max_matches = min(max(integer_value(action_input.get("maxMatches"), 50), 1), 100)
        matches: list[dict[str, Any]] = []
        candidates = (candidate for candidate in root.rglob("*") if candidate.is_file())
        for index, path in enumerate(candidates):
            if index >= max(1, self.max_search_files) or len(matches) >= max_matches:
                break
            if path.stat().st_size > 1_000_000:
                continue
            try:
                lines = path.read_text(encoding="utf-8").splitlines()
            except UnicodeDecodeError:
                continue
            for line_number, line in enumerate(lines, start=1):
                if query in line:
                    matches.append({"path": self._relative(path), "lineNumber": line_number, "line": line})
                    if len(matches) >= max_matches:
                        break
        return {"query": query, "matches": matches}

    def _propose_patch(self, action_input: dict[str, Any]) -> dict[str, Any]:
        path = self._resolve_path(string_value(action_input, "path", ""))
        content = string_value(action_input, "content", "")
        patch = string_value(action_input, "patch", "")
        old_content = path.read_text(encoding="utf-8") if path.is_file() else ""
        rendered_patch = patch or "".join(
            difflib.unified_diff(
                old_content.splitlines(keepends=True),
                content.splitlines(keepends=True),
                fromfile=self._relative(path),
                tofile=self._relative(path),
            )
        )
        return {
            "path": self._relative(path),
            "summary": string_value(action_input, "summary", ""),
            "contentBytes": len(content.encode()),
            "patch": rendered_patch,
            "wouldCreate": not path.exists(),
            "applyAction": "workspace_apply_patch",
        }

    def _apply_patch(self, action_input: dict[str, Any]) -> dict[str, Any]:
        if not self.write_enabled:
            return {"status": "error", "message": "workspace writes are disabled"}
        path = self._resolve_path(string_value(action_input, "path", ""))
        content = string_value(action_input, "content", "")
        patch = string_value(action_input, "patch", "")
        old_content = path.read_text(encoding="utf-8") if path.is_file() else ""
        next_content = apply_unified_diff(old_content, patch) if patch else content
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(next_content, encoding="utf-8")
        return {"status": "written", "path": self._relative(path), "bytes": len(next_content.encode())}

    def _run_shell(self, action_input: dict[str, Any]) -> dict[str, Any]:
        if not self.shell_enabled:
            return {"status": "error", "message": "workspace shell is disabled"}
        command = string_value(action_input, "command", "").split()
        if not self._allowed_command(command):
            return {"status": "error", "message": "command is not allowed"}
        timeout = min(max(integer_value(action_input.get("timeoutSeconds"), self.command_timeout_seconds), 1), 30)
        try:
            # _allowed_command validates the executable and all privileged subcommands.
            completed = subprocess.run(  # nosec B603
                command,
                cwd=self.root,
                capture_output=True,
                timeout=timeout,
                check=False,
            )
        except subprocess.TimeoutExpired:
            return {"status": "error", "message": "command timed out"}
        stdout, stdout_truncated = truncate_bytes(completed.stdout, self.max_command_output_bytes)
        stderr, stderr_truncated = truncate_bytes(completed.stderr, self.max_command_output_bytes)
        stdout_text = stdout.decode("utf-8", errors="replace")
        stderr_text = stderr.decode("utf-8", errors="replace")
        return {
            "exitCode": completed.returncode,
            "stdout": stdout_text,
            "stderr": stderr_text,
            "output": stdout_text + stderr_text,
            "truncated": stdout_truncated or stderr_truncated,
        }

    def _resolve_path(self, raw_path: str) -> Path:
        path = (self.root / (raw_path or ".")).resolve()
        if not path.is_relative_to(self.root):
            raise ValueError("path escapes workspace root")
        return path

    def _relative(self, path: Path) -> str:
        return "." if path == self.root else path.relative_to(self.root).as_posix()

    def _file_summary(self, path: Path) -> dict[str, Any]:
        result: dict[str, Any] = {"path": self._relative(path), "type": "directory" if path.is_dir() else "file"}
        if path.is_file():
            result["size"] = path.stat().st_size
        return result

    def _allowed_command(self, command: list[str]) -> bool:
        if not command or command[0] not in self.allowed_commands:
            return False
        if command[0] == "pwd":
            return len(command) == 1
        if command[0] in {"ls", "rg"}:
            return self._args_within_workspace(command)
        if command[0] == "git":
            return len(command) >= 2 and command[1] in self.allowed_git_subcommands
        return command[0] == "mvn" and "test" in command and all(
            item in {"mvn", "test", "-q"} or item.startswith("-D") for item in command
        )

    def _args_within_workspace(self, command: list[str]) -> bool:
        """Every path-like argument of ls/rg must stay inside the workspace root.

        Mirrors the Java WorkspaceRuntime.argsWithinWorkspace hardening: both
        commands accept path arguments, so without this check an agent could
        read arbitrary host files (e.g. ``rg pattern /etc/passwd``). For rg the
        first non-flag argument is the search pattern, not a path; anything
        that is not unambiguously a workspace path fails closed.
        """
        pattern_pending = command[0] == "rg"
        seen_separator = False
        for item in command[1:]:
            if not seen_separator and item == "--":
                seen_separator = True
                continue
            if not seen_separator and item.startswith("-"):
                continue
            if pattern_pending:
                pattern_pending = False
                continue
            resolved = (self.root / item).resolve()
            if not resolved.is_relative_to(self.root):
                return False
        return True


def apply_unified_diff(original: str, patch: str) -> str:
    if not patch.strip():
        return original
    original_lines = java_diff_lines(original)
    patch_lines = java_diff_lines(patch)
    result: list[str] = []
    source_index = 0
    line_index = 0
    while line_index < len(patch_lines):
        line = patch_lines[line_index]
        if line.startswith(("---", "+++")):
            line_index += 1
            continue
        if not line.startswith("@@"):
            line_index += 1
            continue
        header = line.split(" ")
        old_range = header[1]
        old_start = int(old_range[1:].split(",")[0]) - 1
        result.extend(original_lines[source_index:old_start])
        source_index = old_start
        line_index += 1
        while line_index < len(patch_lines) and not patch_lines[line_index].startswith("@@"):
            change = patch_lines[line_index]
            marker, content = change[:1], change[1:]
            if marker == " ":
                if source_index >= len(original_lines) or original_lines[source_index] != content:
                    raise ValueError("patch context does not match workspace file")
                result.append(original_lines[source_index])
                source_index += 1
            elif marker == "-":
                if source_index >= len(original_lines) or original_lines[source_index] != content:
                    raise ValueError("patch removal does not match workspace file")
                source_index += 1
            elif marker == "+":
                result.append(content)
            elif marker != "\\":
                raise ValueError("invalid unified diff")
            line_index += 1
    result.extend(original_lines[source_index:])
    return "\n".join(result)


def java_diff_lines(value: str) -> list[str]:
    trimmed = value[:-1] if value.endswith("\n") else value
    return trimmed.split("\n") if trimmed else []


def truncate_bytes(value: bytes, limit: int) -> tuple[bytes, bool]:
    size = max(1, limit)
    return (value[:size], len(value) > size)


def string_value(values: dict[str, Any], key: str, fallback: str) -> str:
    value = values.get(key)
    return str(value).strip() if value is not None and str(value).strip() else fallback


def integer_value(value: Any, fallback: int) -> int:
    try:
        return int(value) if value is not None else fallback
    except (TypeError, ValueError):
        return fallback


def elapsed_ms(started: float) -> int:
    return max(0, int((perf_counter() - started) * 1000))
